File: j2kt/jre/java/native/java/util/List.java
/*
 * Copyright 2007 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.util;

import static javaemul.internal.KtNativeUtils.ktNative;

import java.util.function.UnaryOperator;
import javaemul.internal.annotations.KtName;
import javaemul.internal.annotations.KtNative;
import javaemul.internal.annotations.KtOut;
import org.jspecify.annotations.NullMarked;
import org.jspecify.annotations.Nullable;

/**
 * See <a href="https://docs.oracle.com/javase/8/docs/api/java/util/List.html">the official Java API
 * doc</a> for details.
 */
@KtNative(
    name = "kotlin.collections.List",
    mutableName = "kotlin.collections.MutableList",
    bridgeName = "javaemul.lang.MutableListJvm",
    companionName = "java.util.List")
@NullMarked
public interface List<@KtOut E extends @Nullable Object> extends Collection<E> {
  static <E> List<E> copyOf(Collection<? extends E> coll) {
    return ktNative();
  }

  static <E extends @Nullable Object> List<E> of() {
    return ktNative();
  }

  static <E> List<E> of(E e1) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5, E e6) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7, E e8) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7, E e8, E e9) {
    return ktNative();
  }

  static <E> List<E> of(E e1, E e2, E e3, E e4, E e5, E e6, E e7, E e8, E e9, E e10) {
    return ktNative();
  }

  @SafeVarargs
  static <E> List<E> of(E... elements) {
    return ktNative();
  }

  void add(int index, E element);

  boolean addAll(int index, Collection<? extends E> c);

  E get(int index);

  int indexOf(@Nullable Object o);

  int lastIndexOf(@Nullable Object o);

  ListIterator<E> listIterator();

  ListIterator<E> listIterator(int from);

  @KtName("removeAt")
  E remove(int index);

  default void replaceAll(UnaryOperator<E> operator) {
    ktNative();
  }

  E set(int index, E element);

  default void sort(@Nullable Comparator<? super E> c) {
    ktNative();
  }

  default E getFirst() {
    return ktNative();
  }

  default E getLast() {
    return ktNative();
  }

  default MutableList<E> asMutableList() {
    return ktNative();
  }

  List<E> subList(int fromIndex, int toIndex);

  // Override equals() to teach Error Prone that an ArrayList can be equal to a LinkedList.
  @Override
  boolean equals(@Nullable Object o);
}

File: j2kt/jre/java/native/java/util/List.kt
/*
 * Copyright 2022 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 */
package java.util

import java.util.Collections.Companion.unmodifiableList
import java.util.function.UnaryOperator
import kotlin.Comparator as KotlinComparator
import kotlin.collections.List as KotlinList
import kotlin.collections.replaceAll as kotlinReplaceAll

interface List<E> : MutableList<E>, Collection<E> {
  fun replaceAll(operator: UnaryOperator<E>): Unit = kotlinReplaceAll(operator::apply)

  fun sort(c: KotlinComparator<in E>?): Unit =
    if (c == null) {
      @Suppress("UNCHECKED_CAST") sortBy { it as Comparable<Any> }
    } else {
      sortWith(c)
    }

  fun getFirst(): E = if (isEmpty()) throw NoSuchElementException() else get(0)

  fun getLast(): E = if (isEmpty()) throw NoSuchElementException() else get(size - 1)

  companion object {
    fun <E> of(): KotlinList<E> = createInternal<E>()

    fun <E : Any> of(e1: E): KotlinList<E> = createInternal(e1)

    fun <E : Any> of(e1: E, e2: E): KotlinList<E> = createInternal(e1, e2)

    fun <E : Any> of(e1: E, e2: E, e3: E): KotlinList<E> = createInternal(e1, e2, e3)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E): KotlinList<E> = createInternal(e1, e2, e3, e4)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E, e5: E): KotlinList<E> =
      createInternal(e1, e2, e3, e4, e5)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E, e5: E, e6: E): KotlinList<E> =
      createInternal(e1, e2, e3, e4, e5, e6)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E, e5: E, e6: E, e7: E): KotlinList<E> =
      createInternal(e1, e2, e3, e4, e5, e6, e7)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E, e5: E, e6: E, e7: E, e8: E): KotlinList<E> =
      createInternal(e1, e2, e3, e4, e5, e6, e7, e8)

    fun <E : Any> of(e1: E, e2: E, e3: E, e4: E, e5: E, e6: E, e7: E, e8: E, e9: E): KotlinList<E> =
      createInternal(e1, e2, e3, e4, e5, e6, e7, e8, e9)

    fun <E : Any> of(
      e1: E,
      e2: E,
      e3: E,
      e4: E,
      e5: E,
      e6: E,
      e7: E,
      e8: E,
      e9: E,
      e10: E,
    ): KotlinList<E> = createInternal(e1, e2, e3, e4, e5, e6, e7, e8, e9, e10)

    fun <E : Any> of(vararg elements: E): KotlinList<E> = return createInternal(*elements)

    private fun <E> createInternal(vararg elements: E): KotlinList<E> =
      unmodifiableList(listOf(*elements))

    fun <E : Any> copyOf(coll: kotlin.collections.Collection<out E>): KotlinList<E> = buildList {
      for (element in coll) {
        add(Objects.requireNonNull(element))
      }
    }
  }
}
